250 precursor

e2tha-e opened the pull request:
Addresses #250 

Summary of changes:
Some edits to paramToJson in parameter_hunter.js required when parsing an enterprise Pattern Lab instance and a unit test for it.

e2tha-e commented:
@bmuenzenmeyer Some more precursor work for issue #250 . Please review and merge.

e2tha-e commented:
@bmuenzenmeyer 
cc @james-nash 
#292 seems like it's doing parallel, and possibly superior work in paramToJson. Please review, but hold off on merging, while we come to a consensus on what should be merged.

james-nash commented:
Makes sense. If there's anything I can do to help (e.g. if any of my code is unclear) just shout :-)

bmuenzenmeyer commented:
I haven't made time to review either yet - but agree this area of code is undergoing a lot of change and likely one of these two PRs needs to win out or consume the other.   I'll appreciate any degree of review you both can do in that regard in the meantime.

james-nash commented:
Sure. I have a bit of time this afternoon, so I'll take a closer look at the changes in this pull request.

e2tha-e commented:
@james-nash Could you please hold off review of commit 695471c ? I'll be submitting something more up-to-date hopefully later today. Thanks.

james-nash commented:
Sure thing. Just shout whenever it's ready.

e2tha-e commented:
@james-nash @bmuenzenmeyer Ready for review. Notes on the param parsing are in the docblock to paramToJson.

FYI, I added the JSON5 package and replaced the JSON object with it, for better error messaging on malformed JSON. Running `npm install` and `grunt nodeunit` will give a great example of its advantages.

james-nash commented on core/lib/list_item_hunter.js:
```diff
@@ -13,6 +13,7 @@
 var list_item_hunter = function () {
 
   var extend = require('util')._extend,
+    JSON = require('json5'),
```
I wonder if it might be more appropriate to name this `JSON5`, so that people don't mistake it for the built-in `JSON` object.

e2tha-e replied:
@james-nash Had to debate over this myself, and I'm still torn. @bmuenzenmeyer What are your thoughts?

james-nash replied:
I'm not too fussed myself. However, if you were to change it to `JSON5` then I'd suggest doing that throughout all files where you use that library.

bmuenzenmeyer replied:
Forgive my naivety on JSON5 (it looks like a nice solution to a problem we've all faced), but it doesn't sit right with me. My biggest worry is downstream consumption of .json5 files (I'd also argue that things like trailing commas are not good form. Yes comments would be nice). If people get accustomed to writing these files, or perhaps worse, writing them as "invalid" pure .json files, they might shoot themselves or others in the foot. Is this fear overblown? 

Another consideration in my rather scattershot comment is that it looks like they don't follow semver (yet), which *could* create problems.

... more thinking ...

Put simply, can you explain how adding JSON5 makes #250 (which PR references by name) and the parameter_hunter.paramToJson() method better? 

I am an outsider on the specifics of this issue, and have no doubt that a lot of thought has been put into getting us this far, but I need your help in understanding if this is straying too far from the goal of this issue/fix.    https://github.com/pattern-lab/patternlab-node/pull/293/commits/c8acb5ad9aebc9051ab5a2906b14b4fadcd7a55e  mentions better error handling, but in what way is it better than the existing try catch logic? http://codepen.io/bmuenzenmeyer/pen/66eb18de343ab1954abc53537fde09f0 has an example that is descriptive enough, coupled with a bit more logging on PL's part, would seem to do the trick in my mind.

Thoughts?

cc @geoffp

james-nash replied:
It's important to note that the pattern parameter syntax is only vaguely defined here http://patternlab.io/docs/pattern-parameters.html#pattern-parameter-syntax , but since the examples have keys that are not double-quoted, it's definitely not pure JSON. It's JSON-ish :-)

I'd be in favour of keeping our parsing of the pattern parameter syntax quite lax, since it's conceivable that users might copy-paste chunks of JSON data (so then keys might be double-quoted) and, as per issue #145 it's desirable to allow HTML code to passed in as a value so allowing single- as well as double-quoted string values makes sense to me (saves you having to escape every quote around HTML attributes and stuff like that).

The changes in this PR which make use of JSON5 enable all of that, so I think they're worth keeping.

Since, I've been focussed primarily on the parsing of pattern parameters, which is in `parameter_hunter.paramToJson()`, I'm not as familiar with the other code though. Perhaps @e2tha-e can comment on that.

FWIW, where proper JSON data is expected, I'd agree that it might be better to be stricter (or at least display warnings in the build output) so as not to encourage sloppy JSON code that may not be interoperable with other tools or PatternLab versions.

james-nash commented on core/lib/parameter_hunter.js:
```diff
@@ -13,110 +13,227 @@
 var parameter_hunter = function () {
 
   var extend = require('util')._extend,
+    JSON = require('json5'),
```
Same comment as above. I'd consider renaming to `JSON5` for clarity.

james-nash commented on core/lib/parameter_hunter.js:
```diff
@@ -13,110 +13,227 @@
-        quotePos = paramString.search(/[^\\]'/);
+      //check if searching for a key
+      if (paramString[0] === '{' || paramString[0] === ',') {
+        paramString = paramString.substring([1], paramString.length).trim();
```
The first `substring()` arg should be an integer, not an array (probably just a typo). It happens to work since JS coerces `[1]` to `1` in this case, but still worth tidying up to avoid confusion/bugs in future.

e2tha-e replied:
@james-nash Great catch!

james-nash commented on core/lib/parameter_hunter.js:
```diff
@@ -13,110 +13,227 @@
+        //wrapper pairs and conclude that their contents are values
+        switch (paramString[0]) {
+          case '"':
+            regex = /^"(.|\s)*?"/;
```
This regex can be simplified to `/^"(.)*?"/`. The `.` matches any character, whitespace included.
Same goes for the regex below.
I made this change locally on my machine and re-ran the tests to verify that it does indeed work.

e2tha-e replied:
@james-nash `(.)*?` doesn't work with newline characters.

james-nash replied:
@e2tha-e Ah! OK, I missed that. Ignore that comment then.

james-nash commented:
Ok, I've finished reviewing. It looks great!
There's few minor things I commented inline (and I see that you've already responded to them), but otherwise no issues.

I see that you brought across my additional test cases, so this solution successfully fixes the same problems I was attempting to fix in my PR #292. As far as I'm concerned, #292 can be rejected and this PR accepted instead.

e2tha-e commented:
@james-nash Thanks a million for the code review! I just pushed that typo-fix, and so am waiting on further review and merge. So hopefully issue #291 can be put to bed soon!

e2tha-e commented:
@bmuenzenmeyer @james-nash I'll be offline on vacation for a few days, so don't mind the radio silence too much. Looking forward to be back at it when I'm back :smile:

james-nash commented:
@e2tha-e: No worries. Enjoy your vacation!

james-nash commented on core/lib/parameter_hunter.js:
```diff
@@ -13,110 +13,230 @@
-      paramString = paramString.replace(/(^\s*[\{|\,]\s*)'([^']+)'(\s*\:)/, '$1"$2"$3');
+      //check if searching for a key
+      if (paramString[0] === '{' || paramString[0] === ',') {
+        paramString = paramString.substring([1], paramString.length).trim();
```
I found another instance of using an array instead of a number for `substring()`'s first argument. Must have missed that last time around - sorry!

e2tha-e commented:
@bmuenzenmeyer I fixed that typo pointed out by @james-nash . I've also renamed the instances of JSON5 to be clear that it is not the native JSON object. I'm happy to replace JSON5 entirely, but I don't think it detracts from Pattern Lab in any way. I don't think it by default encourages lax JSON in the actual JSON files. Unless you peak under the hood, how would you know? JSON5 does, however, provide much better error handling than native JSON. I initially experimented with JSON5 to see if it would it could help with converting params to JSON. I found that it wasn't necessary, but boy were those error-handling message helpful. As such, I left it in. Any thoughts or suggestions are welcome.

e2tha-e commented:
@bmuenzenmeyer Here's an example of the difference between native JSON and JSON5 exception handling. It's from running parameter_hunter_tests.js:

Native JSON:
```
[SyntaxError: Unexpected token ,]
```
JSON5:
```
{ [SyntaxError: Unexpected ',' at line 1 column 17 of the JSON5 data. Still to read: ",\"\":missing-key,\"\":,"]
  message: 'Unexpected \',\' at line 1 column 17 of the JSON5 data. Still to read: ",\\"\\":missing-key,\\"\\":,"',
  at: 16,
  lineNumber: 1,
  columnNumber: 17 }
```
In issue #250 , @vebersol and I mentioned we worked on enterprise Pattern Lab projects with hundreds of patterns. I know personally, that the "Unexpected token" syntax error was the bane of my existence. There's no way to lint parameters, and how was anyone expected to be able to quickly track down where that error came from?

Again, strictly syntaxed JSON is a requirement for .json files because they are read by fs.readJSONSync(). JSON5 has nothing to do with those files. I also wouldn't mind revisiting JSON5 being a drop-in replacement for JSON in the few instances where it's used, in the same manner that fs-extra is a drop-in replacement for fs.

bmuenzenmeyer commented:
I'm coming around to this.

I made `00-homepage.json` malformed and it silently failed, failing to apply any data to the page from the template. It's because [this](https://github.com/pattern-lab/patternlab-node/blob/c2bf904e52d3e036e11b5ab0f992b9b5ffca2c07/core/lib/pattern_assembler.js#L183). 

![image](https://cloud.githubusercontent.com/assets/298435/14340421/10689034-fc4d-11e5-8ef7-44b5b3fb1fbe.png)

This test is probably outside the scope of the current issue/fix since it relies on `fs` - but that's the sort of error you are striving to make better. Let me give this some more thought

bmuenzenmeyer commented:
I'm realizing [more and more](https://github.com/pattern-lab/patternlab-node/issues/305) that users are taking PL Node into territory and scale I haven't yet experienced personally or professionally. And that there is merit to making the library more robust, even at the expense of some sense of "purity" I define in my head. Your contribution addresses a need, is thoughtfully proposed, thorough in its implementation, and covered with unit tests. What more could one ask for?

I'll need to be careful to forward port this into `dev-2.0-core`

Thanks!
